ibm-java: new package

mwkrentel opened the pull request:
This is a binary package of the IBM Java SDK for big and little-endian
powerpc (power7, 8 and 9).  The jdk and openjdk packages only install
on x86_64.

tgamblin commented on var/spack/repos/builtin/packages/ibm-java/package.py:
```diff
@@ -0,0 +1,69 @@
+        fd.write('INSTALLER_UI=silent\n')
+        fd.write('USER_INSTALL_DIR=%s\n' % prefix)
+        fd.write('LICENSE_ACCEPTED=TRUE\n')
+        fd.close()
```
Two things: does this belong in a patch function?  That way it'll be done once per stage directory instead of on every install.  Not a big deal, but we tend to put modifications like this there so you can see them *before* install if you want to debug.

Second, minor stylistic thing -- why not:

```python
with open('properties', 'w') as fd:
    fd.write('INSTALLER_UI=silent\n')
    fd.write('USER_INSTALL_DIR=%s\n' % prefix)
    fd.write('LICENSE_ACCEPTED=TRUE\n')
```

And if I'm getting *really* technical, that's not an `fd` - fd's are returned by `os.open`; file objects are returned by `open`, but I really don't care if you call it fd.

mwkrentel commented:
It's not patching an existing file.  It's a new file, outside of the
archive.

When I was doing this manually to see how it works, sometimes running
the archive file would overwrite the properties file with the new
values you used for that run.  That's what the directions tell you to
do for a non-interactive install: run it interactively, let it create
a properties file, then run it again with -f.

Maybe I'm being overly cautious, but I wanted to make sure that it
would work from a leftover directory.  There's no problem with a clean
install.  But if there was a leftover file, I wanted to whack it with
the right contents.

It's only 3 lines.  Normally, these would just be command-line
arguments, but that's not how the archive works.

I'll change the write to use with.

mwkrentel commented:
I changed the open to use with.
Ok?

mwkrentel commented:
@tgamblin  Actually, before merging, let me check on a couple points.

1. Should I add ibm-java (and openjdk) to the java providers entry in
`etc/spack/defaults/packages.yaml` ?

2. Is there a way to condition or restrict the providers list by arch
type?  That is, jdk and openjdk are only for x86_64, and ibm-java is
for ppc64 and ppc64le.

3. Should I add a conflicts() entry to the jdk and openjdk packages
for non-x86 arches?

mwkrentel commented:
@tgamblin The more I think about it, the more I think the answers to
my questions are yes.

So, I'd like to add 'ibm-java' and 'openjdk' to the list of java
providers in `/etc/spack/defaults/packages.yaml`.

```
packages:
  all:
    providers:
      java: [jdk, openjdk, ibm-java]
      ...
```

Is that ok?  I want to be extra careful about this.  Touching the
defaults `packages.yaml` file, I feel like I should have top-secret
clearance or something.  :-)

Btw, is there some reason openjdk is not already in the providers list
for java?  Was that just an oversight when the package was added, or
is there some reason it shouldn't be there?

Also, I'd like to add conflicts('target=ppc64') to the jdk and openjdk
packages since they only work on x86_64.  That way, concretization
should skip over them on ppc64.  Ok?

And with the `providers:` list, is there a way to restrict it by arch
or target?  That is, on x86_64, you want `[jdk, openjdk]`, but on
ppc64 and ppc64le, you want `[ibm-java]`.

Is there a syntax for that?

mwkrentel commented:
Ok, I added openjdk and ibm-java to the list of java providers and
added conflicts() for powerpc to jdk and openjdk.

At this point, I think I'm done.
